### ISSUE TITLE
[bug] Build fails with Protobuf 24.x

### ISSUE THREAD

**Issue by tuaris** [dogecoin/dogecoin#3547]
# Bug Report

First I would like to mention that after I have worked around the issue mentioned in the bug report, I am very pleased to report that Dogecoin Core QT compiles and runs "out of the box" on FreeBSD without any patches.  Well done!  If the Dogecoin project can commit to making sure that it continues to work on FreeBSD, unaltered, I will submit this to the FreeBSD ports team so they can include it as part of the official ports tree.

**Current behavior:**
FreeBSD has updated their default version of Protobuf in the ports tree to 24.4,1 (as of this bug report).  Dogecoin fails to build due this new version of Protobuf:

```
/usr/local/include/google/protobuf/port_def.inc:210:1: error: static_assert failed due to requirement '201103L >= 201402L' "Protobuf only supports C++14 and newer."
static_assert(PROTOBUF_CPLUSPLUS_MIN(201402L), "Protobuf only supports C++14 and newer.");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Attempting to force a c++14 compiler (with `CXXFLAGS="-std=c++14"`) solves the above, but results in Dogecoin failing to build at another point:

```
In file included from qt/utilitydialog.cpp:10:
In file included from ./qt/utilitydialog.h:11:
In file included from ./qt/walletmodel.h:8:
In file included from ./qt/paymentrequestplus.h:10:
In file included from ./qt/paymentrequest.pb.h:24:
In file included from /usr/local/include/google/protobuf/io/coded_stream.h:134:
In file included from /usr/local/include/absl/strings/cord.h:78:
In file included from /usr/local/include/absl/container/inlined_vector.h:53:
In file included from /usr/local/include/absl/container/internal/inlined_vector.h:30:
In file included from /usr/local/include/absl/container/internal/compressed_tuple.h:40:
/usr/local/include/absl/utility/utility.h:164:12: error: no member named 'in_place_t' in namespace 'std'
using std::in_place_t;
      ~~~~~^
/usr/local/include/absl/utility/utility.h:165:12: error: no member named 'in_place' in namespace 'std'
using std::in_place;
      ~~~~~^
```

The 'workaround' is to remove the [devel/protobuf](https://www.freshports.org/devel/protobuf/) port which is at version 24.4.1 and replace it with [devel/protobuf3](https://www.freshports.org/devel/protobuf3/).  Then re-attempt to compile using c++11.

Unfortunately the two ports can not co-exist.  I can see that as a potential problem.

**Expected behavior**
Dogecoin Core should build with c++14 and with Google's latest version of Protobuf

**Steps to reproduce:**
- Install FreeBSD 13 or 14
- Checkout ports tree
- Checkout Dogecoin port located at https://github.com/tuaris/FreeBSD-Coin-Ports
- Install `devel/protobuf`
- Attempt to build the port: `cd /path/to/FreeBSD-Coin-Ports/net-p2p/dogecoin; make build`
- Observe error
- Remove `devel/protobuf`
- Install `devel/protobuf3`
- Port build successfully

**Dogecoin Core version**
v1.14.7.0-330d9f5

**Machine specs**
- OS: FreeBSD
- CPU: Intel
- RAM: 32G
- Disk size: 100G
- Disk Type (HD/SDD): SSD

**Extra information**
The code to functional port can be viewed at: https://github.com/tuaris/FreeBSD-Coin-Ports/tree/master/net-p2p/dogecoin

**Comment by patricklodder** [comment]
> ```
> /usr/local/include/absl/utility/utility.h:164:12: error: no member named 'in_place_t' in namespace 'std'
> using std::in_place_t;
>       ~~~~~^
> /usr/local/include/absl/utility/utility.h:165:12: error: no member named 'in_place' in namespace 'std'
> using std::in_place;
>       ~~~~~^
> ```
> [..]
> Dogecoin Core should build with c++14 and with Google's latest version of Protobuf

`std::in_place_t` is c++17, not 14. I think this is caused by how abseil is compiled with the ports package. There's a similar issue on MacOS where [the brew abseil package requires C++17]() ([brew link](https://github.com/Homebrew/homebrew-core/blob/905bb19d6f46c4f61e3793178c95165cc1c5d21b/Formula/a/abseil.rb#L39)). I'm by no means an expert on ports, as I haven't actively used it for a very long time, but looking at it with my limited knowledge I suspect that [the ports package also requires c++17](https://github.com/freebsd/freebsd-ports/blob/a18891a45d2958197e39a52249c1c7516f3a6bf3/devel/abseil/Makefile#L11).

For Dogecoin Core, I was already planning to add an `--enable-c++17` flag for 1.15.0, which we could - depending on impact - backport to a 1.14 patch release.

**Comment by patricklodder** [comment]
@tuaris: This should no longer produce the errors on `in_place` and `in_place_t` anymore in the `1.15.0-dev` branch (at fdf534c13), when you use `--enable-c++17` when configuring. Could you please check if that solved your issue?

**Comment by tuaris** [comment]
@patricklodder your changes work (at least on FreeBSD 13.2)  I did have to make one small patch in regards to miniupnp

https://github.com/tuaris/FreeBSD-Coin-Ports/blob/6f49fef3d9cb20616b3d5c35d3674bafbe21b69d/net-p2p/dogecoin/files/patch-src_net.cpp

I did not run into the issue reported in https://github.com/dogecoin/dogecoin/issues/3673, but I'm using an older version of FreeBSD.  I'll run some additional tests on some of my build boxes.

Btw, how is CI handled?  Would it be possible to incorporate FreeBSD testing?  The ports system makes it super easy, and the most time consuming part of getting setup is already done with this: https://github.com/tuaris/FreeBSD-Coin-Ports/tree/master/net-p2p/dogecoin

**Comment by tuaris** [comment]
> Perhaps it is more useful to do nightly builds against ports managed dependencies? Get early warning in through that?

Yeah, I think that might be the only option.  The FreeBSD ports get updated a lot more frequently and that's usually what breaks builds.  They do cut quarterly branches, but that too may be too often.

Anyway, I did test the build against [13.3](http://pkg.ent.morante.net/poudriere/data/13amd64-deluxe/2024-12-17_22h49m20s/logs/dogecoin-1.14.9.log), [14.1](http://pkg.ent.morante.net/poudriere/data/14amd64-deluxe/2024-12-18_01h49m48s/logs/dogecoin-1.14.9.log), and [14.2](http://pkg.ent.morante.net/poudriere/data/142amd64-deluxe/2024-12-18_01h58m06s/logs/dogecoin-1.14.9.log).  They all worked using the patch in https://github.com/dogecoin/dogecoin/issues/3673.
